src/test/java/jenkins/plugins/git/GitToolChooserTest.java

```diff
@@ -0,0 +1,478 @@
+        String remote = "https://bitbucket.com/rishabhBudhouliya/git-plugin.git";
+        sampleRepo.init();
+        store.addCredentials(Domain.global(), createCredential(CredentialsScope.GLOBAL, "github"));
+        store.save();
+        buildAProject(sampleRepo, false);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(remote, list.get(0), "github", gitExe);
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    In the event of having an extension which is applicable to the remote URL but throws an exception due to some
+    reason from the implemented git provider plugin, the estimator handles the exception by silently logging an
+    "INFO" message and returns no recommendation.
+     */
+    @Test
+    public void testSizeEstimationWithException() throws Exception {
+        String remote = "https://bitbucket.com/rishabhBudhouliya/git-plugin.git";
+        sampleRepo.init();
+        store.addCredentials(Domain.global(), createCredential(CredentialsScope.GLOBAL, "github"));
+        store.save();
+        buildAProject(sampleRepo, false);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(remote, list.get(0), "github", gitExe);
+
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    In case of having no user credentials, the git plugin expects the `implementers` of the extension point to handle
+    and try querying for size of repo, if it throws an exception we catch it and recommend "NONE", i.e, no recommendation.
+     */
+    @Test
+    public void testSizeEstimationWithNoCredentials() throws Exception {
+        sampleRepo.init();
+
+        buildAProject(sampleRepo, true);
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed by user and git is present in the machine
+        String gitExe = "git";
+
+        GitToolChooser sizeEstimator = new GitToolChooser(sampleRepo.toString(), list.get(0), null, gitExe);
+
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    /*
+    Tests related to git tool resolution
+    Scenario 1: Size of repo is < 5 MiB, "jgit" should be recommended
+     */
+    @Test
+    public void testGitToolChooserWithLessThan5Mb() throws Exception {
+        String url = "https:github.com";
+        Item context = Mockito.mock(Item.class);
+        String credentialsId = null;
+
+        // With JGit, we don't ask the name and home of the tool
+        GitTool tool = new GitTool("my-git", "/usr/bin/git", Collections.<ToolProperty<?>>emptyList());
```

Comment:
Interesting... will look but was wondering if `ToolProperty` is something that could demonstrate capabilities of a `GitTool`?

Reply:
Will have to look at how ToolProperty is used, I haven't checked that either. Thanks for raising this concern!

Reply:
May not bear any fruit. Just a thought 😃